src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -157,39 +162,179 @@ internal void ClearCaches()
+                // implementations of an interface that is supported by the source generator.
+                // IMPORTANT: nearest-ancestor resolution should be reserved for weakly-typed serialization.
+                // Attempting to use it in strongly typed operations or deserialization will invariably
+                // result in an invalid cast exception, so use with caution.
+
+                Debug.Assert(!entry.HasResult);
+                CacheEntry? candidate = null;
+                Type? candidateType = null;
+
+                for (Type? current = type.BaseType; current != null; current = current.BaseType)
+                {
+                    if (current == JsonTypeInfo.ObjectType)
+                    {
+                        // Avoid falling back to the contract for object since it's polymorphic
+                        // and it would try to send us back to the runtime type that isn't supported.
+                        break;
+                    }
+
+                    candidate = GetOrAddCacheEntry(current);
+                    if (candidate.HasResult)
+                    {
+                        // We found a type in the class hierarchy that has a contract -- stop looking further up.
+                        candidateType = current;
+                        break;
+                    }
+                }
+
+                foreach (Type interfaceType in type.GetInterfaces())
+                {
+                    CacheEntry interfaceEntry = GetOrAddCacheEntry(interfaceType);
+                    if (interfaceEntry.HasResult)
+                    {
+                        if (candidateType != null)
+                        {
+                            if (interfaceType.IsAssignableFrom(candidateType))
+                            {
+                                // The previous candidate is more derived than the
+                                // current interface -- keep our previous choice.
+                                continue;
+                            }
+                            else if (candidateType.IsAssignableFrom(interfaceType))
+                            {
+                                // The current interface is more derived than the
+                                // previous candidate -- replace the candidate value.
+                            }
+                            else
+                            {
+                                // We have found two possible ancestors that are not in subtype relationship.
+                                // This indicates we have encountered a diamond ambiguity -- abort search and record an exception.
+                                NotSupportedException nse = ThrowHelper.GetNotSupportedException_AmbiguousMetadataForType(type, candidateType, interfaceType);
+                                candidate = new CacheEntry(ExceptionDispatchInfo.Capture(nse));
```

Comment:
Note that as this exception wasn't thrown, there isn't going to be anything "captured" for it in terms of a stack trace.  Not sure if that matters for your needs.